xsdgen/xsdgen.go

```diff
@@ -810,14 +810,18 @@ func (cfg *Config) genComplexTypeMethods(t *xsd.ComplexType, overrides []fieldOv
 				{{end}}
 			}
 			overlay.T = (*T)(t)
+			err := d.DecodeElement(&overlay, &start)
+			if err != nil {
+				return err
+			}
 			{{range .Overrides}}
-			overlay.{{.FieldName}} = (*{{.ToType}})({{if not .IsPtr}}&{{end}}overlay.T.{{.FieldName}})
+			overlay.T.{{.FieldName}} = ({{if .IsPtr}}*{{end}}{{.FromType}})({{if not .IsPtr}}*{{end}}overlay.{{.FieldName}})
 			{{if .DefaultValue -}}
 			// overlay.{{.FieldName}} = {{.DefaultValue}}
 			{{end -}}
 			{{end}}
```

Comment:
What is happening here?

Reply:
This is where the _magic_ happens! So, this is the actual fix. The string here is a big template that is used when generating "complex types" from an xsd. Previously, it would've ended up looking something like:

```go
var overlay struct {
    *T
     Firstrevisiondate  *xsdDate `xml:"urn:books first_revision_date,omitempty"`
}
overlay.T = (*T)(t)
overlay.Firstrevisiondate = (*xsdDate)(&overlay.T.Firstrevisiondate)
return d.DecodeElement(&overlay, &start)
```

Normally, this generated code works exactly how it should. The idea seems to be that `overlay.FirstRevision` points to the same address as `overlay.T.FirstRevision`, so when `overlay` is passed into `d.DecodeElement` and `*xsdDate.UnmarshalText` gets called to update the value for `overlay.FirstRevision`, it will also update the embedded value in `overlay.T.FirstRevision` since that is actually pointing to the same address.

The issue here seems to be that for a lot of our date fields in Workday, we generate the `*time.Time` type which doesn't get initialized as anything (starts out as `nil`), so when we try to point `overlay.FirstRevision` to the address for `overlay.T.FirstRevision`, we're not actually pointing it at anything because there is basically no address to point it too (this is a little oversimplified, but the basic gist should be right).

Here is an example in the playground that illustrates some of what I'm talking about: https://go.dev/play/p/Jy_Gzf2pQRk

Pay attention to the memory addresses and how they change. Specifically, you can see that when the customer unmarshaler runs in the first example where we don't initialize the pointers that the receiver suddenly has a memory address even though the TimeField initially doesn't have an address. This is because the xml decoder has logic that when it reflects over the fields in the struct that basically says "if this is a nil pointer, go ahead and initialize it before running any unmarshaling logic". This is basically where the bug was coming from. When it does this, the receiver is now pointing at the newly created pointers address, which means when you update the value of that pointer you're not updating the original `overlay.T.FirstRevision` field since they don't point at the same address.

With my change to the template, the generated code now looks like:


```go
	overlay.T = (*T)(t)
	err := d.DecodeElement(&overlay, &start)
	if err != nil {
		return err
	}
	overlay.T.Firstrevisiondate = (*time.Time)(overlay.Firstrevisiondate)
```

The primary difference here is that instead of relying on the right variables getting set by updating a different variable that hopefully points at the same memory address, we're explicitly updating the field that we want to set (`overlay.T.FirstRevision`).

This is a long explanation and I'm still only 95% sure that I truly know what is going on, but if you want to schedule some time to go over this more we absolutely can!